go/mysql/conn.go

```diff
@@ -464,14 +472,13 @@ func (c *Conn) startEphemeralPacket(length int) []byte {
 }
 
 // writeEphemeralPacket writes the packet that was allocated by
-// startEphemeralPacket. If 'direct' is set, we write to the
-// underlying connection directly, by-passing the write buffer.
-func (c *Conn) writeEphemeralPacket(direct bool) error {
+// startEphemeralPacket.
+func (c *Conn) writeEphemeralPacket() error {
 	defer c.recycleWritePacket()
 
-	var w io.Writer = c.writer
-	if direct {
-		w = c.conn
+	var w io.Writer = c.conn
+	if c.writer != nil {
```

Comment:
i'd almost like to see this swapout happen with `startBuffering` call, and then instead of calling `flush` directly, `startBuffering` returns a finisher function `func()` that resets c.conn to the direct connectin appropriately and then calls `flush`.

This seems like a lot of trickiness, but I think what I'm trying to go for is to completely abstract away whether or not the connection is buffered or not from the `writeXXX` functions. @LK4D4 initially did this by pushing the buffering down into the `writer` object, and what you're doinig here is actually pulling the buffering control UP into the caller. i think the advantages of making `flush` easier to reason about and see make this approach worthwhile, in which case i'd like to go the rest of the way and get rid of the buffered writer juggling down here.